[WTF][WPE] Read the log level string from an Android system property

aperezdc opened the pull request:
#### f8b520c6e9b1bc98f20ff286f01ff6b1428ec2a7
<pre>
[WTF][WPE] Read the log level string from an Android system property
<a href="https://bugs.webkit.org/show_bug.cgi?id=295175">https://bugs.webkit.org/show_bug.cgi?id=295175</a>

Reviewed by Nikolas Zimmermann and Michael Catanzaro.

Read the value of the debug.WPEWebKit.log system property to determine
the value returned by WTF::logLevelString() to configure logging
channels on Android.

This allows using the &quot;setprop&quot; command line tool to configure logging:

  adb shell setprop debug.WPEWebKit.log &apos;Scrolling,Loading&apos;

* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/android/LoggingAndroid.cpp: Copied from Source/WTF/wtf/unix/LoggingUnix.cpp.
(WTF::logLevelString):
* Source/WTF/wtf/unix/LoggingUnix.cpp: Guard with !OS(ANDROID); remove
unneeded &lt;string.h&gt; header inclusion, and make logLevel &quot;const&quot; as
drive-by fixes.
(WTF::logLevelString):

Canonical link: <a href="https://commits.webkit.org/297467@main">https://commits.webkit.org/297467@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44383f0793c237292487afd19c147c4677e3bec6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111896 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22036 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117914 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62124 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40140 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85003 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35693 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25747 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65443 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18830 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61758 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104384 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95126 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/18901 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121196 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110455 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38923 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28945 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93870 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39304 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96952 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93694 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23906 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38880 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16667 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34940 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38821 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/44316 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134716 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38458 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36263 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41783 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WTF/wtf/unix/LoggingUnix.cpp:
```diff
@@ -29,11 +29,34 @@
+
+static const char* androidLogLevelString()
+{
+    const auto* propertyInfo = __system_property_find("debug.wpewebkit.log");
```
Maybe remove `wpe` here?

spenap replied:
I'm OK with both approaches. 
- `debug.webkit.log` is consistent with `WEBKIT_DEBUG` as we normally use it, 
- `debug.wpewebkit.log` is possibly fairer considering we're only having active Android efforts for the WPEWebKit port

aperezdc replied:
My reason to use `debug.wpewebkit.log` here was that [people have been exploring how to run WebKitGTK on Android](https://gitlab.gnome.org/GNOME/glib/-/issues/3449#note_2473318) (no success yet), so at some point we may want to have also `debug.webkitgtk.log` as well&mdash;we may want to allow configuring logging separately for each port, in case the GTK port is eventually made to work on Android.

Thinking about that, probably the best approach could be using something like this:

```suggestion
    const auto* propertyInfo =
        __system_property_find("debug." LOG_CHANNEL_WEBKIT_SUBSYSTEM ".log");
```

This would result in `debug.WPEWebKit.log` for the WPE port, and `debug.WebKitGTK.log` for the GTK one. Also, this makes the name match the “log tag” passed to `__android_log_{print,vprint,write}` in https://github.com/WebKit/WebKit/pull/47339, making everything a bit more consistent.

Thoughts?

spenap replied:
I think this is a good idea!

aperezdc replied:
I have updated the patch in this way. PTAL.

spenap reviewed (commented):
Philippe has a valid question, but I'm OK with both your approach and his suggestion. The changes look good!

mcatanzaro commented on Source/WTF/wtf/unix/LoggingUnix.cpp:
```diff
@@ -29,11 +29,34 @@
+}
+#endif // OS(ANDROID)
+
 namespace WTF {
```
Here it doesn't actually make any difference since the function is static, but namespace should be opened before `androidLogLevelString`

mcatanzaro reviewed (approved):
I would do this in a separate LoggingAndroid.cpp, because now half of LoggingUnix.cpp is unrelated to Unix. But whatever.

nikolaszimmermann commented on Source/WTF/wtf/android/LoggingAndroid.cpp:
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2015 Igalia S.L.
```
2015? :-)

aperezdc replied:
Heh, sloppy finger there, should have been 2025. I'll fix this before landing.

nikolaszimmermann commented on Source/WTF/wtf/android/LoggingAndroid.cpp:
```diff
@@ -0,0 +1,55 @@
+    const char* propertyValue = nullptr;
+
+    if (const auto* propertyInfo = __system_property_find("debug." LOG_CHANNEL_WEBKIT_SUBSYSTEM ".log")) {
+        __system_property_read_callback(propertyInfo, [](void *userdata, const char*, const char* value, unsigned) {
```
void* userData

aperezdc replied:
I will also fix this before landing.

svillar reviewed (commented):
Shouldn't we include the adb command in docs? It's quite useful for developing and it'd be a shame forcing devs to look in the code if they can just see it on a README

aperezdc committed:
[WTF][WPE] Read the log level string from an Android system property
https://bugs.webkit.org/show_bug.cgi?id=295175

Reviewed by Nikolas Zimmermann and Michael Catanzaro.

Read the value of the debug.WPEWebKit.log system property to determine
the value returned by WTF::logLevelString() to configure logging
channels on Android.

This allows using the "setprop" command line tool to configure logging:

  adb shell setprop debug.WPEWebKit.log 'Scrolling,Loading'

* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/android/LoggingAndroid.cpp: Copied from Source/WTF/wtf/unix/LoggingUnix.cpp.
(WTF::logLevelString):
* Source/WTF/wtf/unix/LoggingUnix.cpp: Guard with !OS(ANDROID); remove
unneeded <string.h> header inclusion, and make logLevel "const" as
drive-by fixes.
(WTF::logLevelString):

Canonical link: https://commits.webkit.org/297467@main
